OIDC Support

pranav-super opened the pull request:
`___REQUIRES_GATEWAY_PR___="131"`
`___REQUIRES_AERIE_PR___="1838"`
[Gateway PR Link](https://github.com/NASA-AMMOS/plandev-gateway/pull/131/)
[Backend PR Link](https://github.com/NASA-AMMOS/plandev/pull/1838/)

## Summary

Adds OpenID Connect (OIDC) authentication to plandev-ui, alongside the existing username/password (JWT) and SSO/CAM login modes. Operators authenticate through their organization's identity provider (e.g., Keycloak) using the standard Authorization Code + PKCE flow, and the UI handles refresh, role-switching, and logout end-to-end. OIDC is the standard protocol for organizational single sign-on and is supported by most modern identity providers (Keycloak, Okta, Azure AD, etc.). This PR enables PlanDev be deployed in environments where the mission's team manages access centrally through their existing OIDC infrastructure rather than maintaining a per-app user store. OIDC is off by default and opt-in with `PUBLIC_AUTH_OIDC_ENABLED=true`.

## Glossary

- **OIDC** — OpenID Connect. A thin identity layer on top of OAuth 2.0. Adds the concept of an "ID token" (proof of who the user is) on top of OAuth's access tokens (what the user is allowed to do).
- **PKCE** — Proof Key for Code Exchange. The standard hardening for OAuth's "authorization code" flow when the client (a browser-loaded SvelteKit app, in our case) can't safely hold a client secret. The browser generates a one-time random `verifier`, sends only its SHA-256 hash to the IdP up front, and reveals the verifier only when exchanging the code for tokens. Prevents an attacker who intercepts the authorization code from being able to redeem it.
- **IdP** — Identity Provider. The system that authenticates users and issues tokens (e.g., Keycloak, Okta, Azure AD). plandev defers all "who is this user" decisions to the IdP.
- **JWKS** — JSON Web Key Set. A public endpoint the IdP exposes (`/protocol/openid-connect/certs` on Keycloak) listing the public keys it uses to sign tokens. Anyone validating a token fetches the matching public key from this URL. This is how plandev-ui, plandev-gateway, and Hasura all independently verify JWT signatures without sharing a secret with the IdP.
- **Access token / ID token / refresh token** — the three JWTs the IdP issues. Access token authorizes API calls; ID token carries identity claims; refresh token is used to get fresh access/ID tokens without re-prompting the user.

## OIDC alongside existing auth modes

PlanDev now supports three mutually exclusive auth modes, selected at runtime via env vars:

| Mode | Selector | Token issuer | User provisioning |
|---|---|---|---|
| JWT | (default) | gateway, signs HS256 from username/password against gateway DB | `gateway login()` → direct DB write |
| SSO/CAM | `PUBLIC_AUTH_SSO_ENABLED=true` | gateway, after validating an external CAM session cookie | `gateway login()` → direct DB write |
| OIDC | `PUBLIC_AUTH_OIDC_ENABLED=true` | external IdP, signs RS256 | `gateway session()` → direct DB write (new, this PR) |

The three are mutually exclusive at runtime — Hasura validates JWTs with one configuration (HS256 *or* RS256+JWKS) and can't accept both simultaneously on the current Hasura v2.12.1.

The auth-mode router lives in `src/hooks.server.ts` and dispatches to `handleJWTAuth`, `handleSSOAuth`, or `handleOIDCAuth`. Each ends with `event.locals.user` populated; downstream code is mode-agnostic.

### Architecture: how the pieces talk in OIDC mode

Full first-time login flow:

```mermaid
sequenceDiagram
    autonumber
    participant B as Browser
    participant U as plandev-ui<br/>(SvelteKit server)
    participant I as IdP<br/>(Keycloak)
    participant G as plandev-gateway
    participant H as Hasura
    participant P as Postgres

    Note over B,U: User clicks "Login Using OIDC"
    B->>U: GET /oidc/login
    U->>U: Generate PKCE verifier, state, nonce<br/>code_challenge = S256(verifier)<br/>(verifier/state/nonce set as short-lived cookies)
    U-->>B: 302 → IdP /authorize<br/>(code_challenge, state, nonce in URL)
    B->>I: GET /authorize (code_challenge, state, nonce)
    Note over B,I: IdP renders login page →<br/>user authenticates (+ consent)
    I-->>B: 302 → /oidc/callback?code=...&state=...<br/>(only after successful login)
    B->>U: GET /oidc/callback?code=...&state=...
    U->>I: POST /token (code + PKCE verifier)
    I-->>U: access_token, id_token, refresh_token
    U->>U: Verify signatures via JWKS<br/>(issuer, audience, nonce)<br/>cached keys — hits IdP only on cache miss
    U-->>B: Set-Cookie + 302 → /plans

    Note over B,P: Per UI server request (page navigation / SSR)
    B->>U: GET /plans
    U->>U: Verify JWT (cached JWKS)
    U->>G: GET /auth/session (Bearer token)
    G->>G: Verify JWT (cached JWKS) + extract claims
    G->>P: SELECT users_and_roles →<br/>INSERT only if new user<br/>(existing users unchanged — no role update/prune)
    G-->>U: { success: true }
    U-->>B: SSR'd page

    Note over B,H: Per GraphQL request — direct to Hasura, no gateway
    B->>H: GraphQL query/subscribe (Bearer token)
    H->>H: Verify JWT (cached JWKS)
    H->>P: data
    P-->>H: rows
    H-->>B: data
```

Notes:

- The **IdP's JWKS endpoint** is the single source of truth for signature verification. plandev-ui, plandev-gateway, **and** Hasura all fetch keys from it independently — none of them share a secret with the IdP or with each other.
- The **gateway is NOT a network gateway** despite the name. The browser hits Hasura, action-server, and workspace-server directly. The gateway only sits in the `/auth/session` path — called once per server request to validate the JWT and (new in this PR) provision first-time users.
- **First-time user provisioning** is step 15 in the diagram (`gateway → SELECT/INSERT users_and_roles`). The gateway writes the row directly via SQL with elevated DB credentials — same pattern JWT/SSO modes use today. It does not update roles for returning users; role changes ride on the JWT.
- **On roles & revocation:** OIDC roles are authoritative in the JWT, **not** in Postgres. Step 15 only provisions a `permissions.users` row for FK integrity on first sight — it never updates or prunes an existing user's roles (unlike CAM/SSO, which does a full `syncRolesToDB` DELETE+upsert). Revoking a role at the IdP takes effect on the **next token** (refresh ~10s before `exp`, or re-login): the new token omits the role, and both Hasura (`x-hasura-allowed-roles` check) and the UI role-switcher honor the token. The DB `allowed_roles` may go stale but is not consulted for OIDC authz.
- **On JWKS:** every "Verify JWT (JWKS)" step (UI, gateway, Hasura) uses an **in-process cached** key set (24h TTL, ~100 keys, rate-limited to 10 fetches/min). A `kid` maps to an immutable key, so verification is local on cache hit; the IdP's `/certs` endpoint is contacted only on a cache miss (cold start or key rotation).


Refresh has its own flow, triggered by a browser-side timer rather than the IdP:

```mermaid
sequenceDiagram
    autonumber
    participant T as Browser timer
    participant B as Browser
    participant U as plandev-ui
    participant I as IdP
    participant H as Hasura (WS)

    Note over T: Set to fire ~10s before<br/>access token's exp claim
    T->>B: setTimeout fires
    B->>U: POST /oidc/refresh
    U->>I: POST /token (refresh_token grant)
    I-->>U: new access/id/refresh tokens
    U->>U: Verify new tokens via JWKS
    U-->>B: Set-Cookie (new tokens)
    B->>B: cookieStore change → restartSharedClient()
    B->>H: close WS (code 4999 "Client Restart") + reconnect<br/>with new accessToken in connectionParams
    B->>B: subscribe(accessTokenDecoded) fires<br/>schedule next refresh for new exp
```

### Why OIDC lives in the UI rather than the gateway

Two pieces of OIDC have to live in the UI no matter what:

1. **PKCE bookkeeping cookies** (`verifier`, `oidc_state`, `oidc_nonce`) must land on the UI's origin because the IdP redirects back to the UI's `/oidc/callback`. The gateway runs on a different origin (e.g., `localhost:9000` vs the UI's `localhost:3000`, and different hosts/ports in production), so a cookie set by the gateway wouldn't be sent on the IdP→UI redirect. They have to live on the UI's origin to be available when the callback handler reads them.
2. **Refresh trigger** has to live in browser JS because the browser hits Hasura **directly** for both HTTP queries and WebSocket subscriptions. The gateway never sees those requests so it can't intercept a 401 from Hasura and refresh on the user's behalf. A setTimeout in browser JS is the only place we can put "refresh before the access token expires."

What we *did* move gateway-side is what fits gateway-side naturally: **user provisioning** (writing `permissions.users` on first login). That matches what JWT/SSO modes already do today via `gateway login() → getUserRoles()`. It uses direct DB writes with elevated credentials, bypassing Hasura's per-role insert permissions.

Responsibilities of Services:

- **UI server**: PKCE handshake, token exchange, JWKS verification, cookie management, refresh scheduling, WS restart-on-refresh, IdP-initiated logout
- **Gateway**: JWT validation on `/auth/session`, lazy user-row upsert via SQL
- **Hasura**: independent JWT validation per request via JWKS

### How API users fit in (aerie-cli, scripts, automation)

OIDC support in this PR is **for the browser flow only**. API users — `aerie-cli`, scheduled jobs, custom scripts hitting Hasura directly — wouldn't use any of this code path. They authenticate against the IdP separately using flows appropriate for non-browser clients (device code flow for interactive CLI logins, client credentials grant for service-to-service automation, etc.) and end up holding a JWT they can send to Hasura/gateway/action-server as `Authorization: Bearer <token>`.

That works *because* Hasura validates every JWT against the IdP's JWKS regardless of where the token came from. So putting browser-side OIDC logic in the UI doesn't restrict API access in any way — `aerie-cli` and other consumers go directly to the IdP. The UI's OIDC code is a browser convenience layer, not a chokepoint.

When `aerie-cli` adds OIDC support, it can use this PR's realm config as a reference for the IdP-side setup (test users, role mappers, default-role attribute) but its client-side code will be different (no PKCE/cookies; device code flow + token storage on disk).

## Technical breakdown

**OIDC flow (UI server-side)**

- `src/lib/server/oidc.ts` — `Client` singleton (PKCE), JWT verification via JWKS, lazy `init()` from `OIDC_WELL_KNOWN_URL`, signature + audience + issuer + nonce checks per OIDC spec.
- `src/routes/oidc/login/+page.server.ts` — Generates PKCE verifier, state, nonce; validates the `back` query param against open-redirect attacks.
- `src/routes/oidc/callback/+page.server.ts` — Validates state, exchanges code for tokens, verifies nonce, sets cookies, redirects back to the original page.
- `src/routes/oidc/refresh/+server.ts` — Refreshes access/id tokens; returns 401 on refresh-token rejection so the client can detect and log out.
- `src/routes/oidc/logout/+server.ts` — Clears local cookies, redirects to IdP logout with `id_token_hint`.

**OIDC flow (UI client-side)**

- `src/lib/stores/oidc.ts` — Refresh scheduling derived from the access token's `exp` claim; cookie-store listener proactively restarts the GraphQL WS on accessToken change so Hasura doesn't unilaterally close the connection on expiry.
- `src/utilities/auth.ts` — Shared `computeRolesFromJWT` (used by all three auth modes).
- `src/lib/types/oidc.ts` — Shared `extractClaims` helper + `ClaimsConfig` type (server and client agree on the JWT claim shape).

**Hooks + layout**

- `src/hooks.server.ts` — Three-mode dispatch (JWT/SSO/OIDC), CSP report-only headers, prefix-anchored "non-protected path" detection.
- `src/routes/+layout.server.ts` — Auth-gate via `enforce(locals.user, userIsDefined)`; prefix-anchored to avoid substring-match false negatives.
- `src/routes/+layout.svelte` — Reactive user store; OIDC-only cookie-store + refresh wiring.

**Gateway-side (cross-repo)**

- `src/packages/auth/functions.ts` — `session()` now calls `getUserRoles` (lazy upsert) so OIDC users get a `permissions.users` row on first sight via direct DB write. Same pattern as JWT/SSO login uses today; bypasses Hasura GraphQL permissions entirely.

## User-visible impacts in OIDC mode

What an operator notices that's different from JWT/SSO mode:

- Login button on `/login` says "Login Using OIDC" and routes through the IdP.
- After IdP auth, no extra "first-time setup" step — gateway provisions the user row on first session check.
- Role switcher in the Nav lists *only* roles the user was granted in the IdP. The list is sorted alphabetically (deterministic across logins). Single-role users see no switcher at all.
- Default role on login comes from the IdP user's `default_role` attribute (per-user, mission-configurable).
- Session expiry: refresh is automatic until the refresh token's IdP lifespan runs out, then logout → `/login?reason=Session expired - please log in again`. The reason survives the IdP roundtrip via a short-lived server-side `logoutReason` cookie set by `/oidc/logout` and consumed by `+layout.server.ts` when it redirects to `/login`. WebSocket subscriptions don't blink during a refresh — the proactive restart cycles cleanly with the new token, and the "Reconnecting..." banner is explicitly suppressed for intentional restarts (gated on close code `4999` + reason `"Client Restart"`, so a foreign close at the same code can't accidentally swallow the banner).
- IdP-initiated logout: clicking logout sends `id_token_hint` to the IdP's logout endpoint so the IdP session is also destroyed (not just the local plandev session).

## How to run

### plandev-ui env vars

Required in plandev-ui's `.env` (or sourced from `.env.test.oidc` for local OIDC test runs):

- **`PUBLIC_AUTH_OIDC_ENABLED`** — set to `true` to enable OIDC. The login page renders the "Login Using OIDC" button when this is set; otherwise the UI falls back to JWT/SSO.
- **`OIDC_CLIENT_ID`** — the ID the IdP assigned to your client (often something like `aerie`). The PKCE flow sends this on every IdP call.
- **`OIDC_REDIRECT_URI`** — where the IdP sends the browser back after auth. Must match what's registered with the IdP. Locally: `http://localhost:3000/oidc/callback`.
- **`OIDC_ISSUER`** — the IdP's issuer URL (e.g., `http://localhost:8000/realms/aerie-dev` for the test realm). Used to validate the `iss` claim on incoming JWTs.
- **`OIDC_JWKS_URL`** — the IdP's JWKS endpoint (`/protocol/openid-connect/certs` on Keycloak). The UI fetches public keys from here to verify token signatures.
- **`OIDC_AUDIENCE`** — the expected `aud` claim on the ID token. Typically equals `OIDC_CLIENT_ID`. Validated only on the ID token per OIDC spec; access tokens are treated as opaque.
- **`OIDC_SCOPES`** — space-separated scopes requested at login. `openid profile email` is the typical baseline; defaults to that if unset.

Either supply the well-known discovery URL OR each endpoint individually:

- **`OIDC_WELL_KNOWN_URL`** — the IdP's well-known config URL (usually contains `/.well-known/openid-configuration`). When set, `OIDC_AUTHORIZATION_URL`, `OIDC_TOKEN_URL`, and `OIDC_LOGOUT_URL` are auto-discovered at startup.
- **`OIDC_AUTHORIZATION_URL`** / **`OIDC_TOKEN_URL`** / **`OIDC_LOGOUT_URL`** — explicit endpoints. Required only if well-known discovery is unavailable.

Optional advanced configuration:

- **`OIDC_ALGORITHMS`** — space-separated list of allowed JWT signing algorithms. Defaults to `RS256`, which is what every modern IdP uses. Override only if your IdP signs with something unusual (`RS384`, `RS512`, `ES256`, etc.).
- **`OIDC_CLAIMS_NAMESPACE`** / **`OIDC_CLAIMS_USER_ID`** / **`OIDC_CLAIMS_ALLOWED_ROLES`** / **`OIDC_CLAIMS_DEFAULT_ROLE`** — claim paths within the JWT. Default to the Hasura convention (`https://hasura.io/jwt/claims` + `x-hasura-user-id` etc.). Override if your IdP can't be configured to emit Hasura-style claims and you map them in Hasura's `claims_map` instead.
- **`PUBLIC_OIDC_CLAIMS_*`** — same four claim paths, but read by browser-side code. Must match the non-PUBLIC versions; the duplication is needed because SvelteKit's `$env/dynamic/public` only exposes `PUBLIC_`-prefixed vars to the browser.
- **`OIDC_CLIENT_SECRET`** — not used by the current PKCE flow (PKCE is what replaces the need for a client secret in public/browser clients). Present in `.env` for forward compatibility if a future flow (e.g., client credentials) is added.

### Hasura + gateway env vars

If using OIDC, PlanDev's `HASURA_GRAPHQL_JWT_SECRET` needs to change from HS256 to RS256 with `jwk_url` matching `OIDC_JWKS_URL`. Example:

```json
{
  "type": "RS256",
  "jwk_url": "http://aerie_keycloak:8000/realms/aerie-dev/protocol/openid-connect/certs",
  "claims_namespace": "https://hasura.io/jwt/claims"
}
```

The same `HASURA_GRAPHQL_JWT_SECRET` is consumed by plandev-gateway, action-server, and workspace-server — all of them need to validate IdP-issued JWTs against the same JWKS.

For local development against the bundled Keycloak test realm:

**Prerequisites**
- An **OIDC-capable gateway image**. The published `aerie-gateway:develop`/`:latest`
  images are HS256-only — OIDC support lives in [plandev-gateway PR #131](https://github.com/NASA-AMMOS/plandev-gateway/pull/131/). Pick one:
  - **Reuse CI's published image (no build):** pass `GATEWAY_IMAGE_TAG=pr-131` to setup
    (the image PR #131 publishes via its `publish` label).
  - **Build it yourself, tagged as the default `oidc-local`:**
    ```sh
    git clone https://github.com/NASA-AMMOS/aerie-gateway.git
    cd aerie-gateway && git checkout feature/oidc-support
    npm ci && npm run build
    docker build -t ghcr.io/nasa-ammos/aerie-gateway:oidc-local .
    cd -
    ```

**Bring up the stack and start the UI**
```sh
# 1. Bring up the OIDC stack (Keycloak + Hasura on RS256 + gateway + everything else).
#    Use the published PR image, or omit GATEWAY_IMAGE_TAG to use a locally-built `oidc-local`.
GATEWAY_IMAGE_TAG=pr-131 npm run test:e2e:oidc:setup

# 2. Start the UI with OIDC env vars sourced, then log in at http://localhost:3000
set -a && . ./.env.test.oidc && set +a && npm run dev
```

### Configuring your IdP (production)

Every consumer — the UI (`extractClaims`), the gateway (`session()`), and Hasura — expects a **nested Hasura claims object** on the token used as the `Authorization: Bearer` credential:

```json
{
  "iss": "https://your-idp/…", "aud": "aerie", "sub": "…", "exp": …,
  "https://hasura.io/jwt/claims": {
    "x-hasura-user-id":       "<stable user id>",
    "x-hasura-default-role":  "viewer",
    "x-hasura-allowed-roles": ["aerie_admin", "user", "viewer"]
  }
}
```

- `x-hasura-default-role` is a **single string**; `x-hasura-allowed-roles` is a **JSON array of strings**.
- The claim key/namespace is renamable via `OIDC_CLAIMS_*` (+ the `PUBLIC_OIDC_CLAIMS_*` browser mirror), but the value **must be a nested object** — `extractClaims` throws otherwise. Aerie has **no flat-claims mode today** (see "IdP compatibility" below).
- Set `HASURA_GRAPHQL_JWT_SECRET` to RS256 + `jwk_url` (shared by Hasura, gateway, action-server, workspace-server).

#### IdP compatibility: the one axis that matters

The only question that decides how much work an IdP is: **can it emit the nested `https://hasura.io/jwt/claims` object natively?**

- **Camp A — emits the nested object natively (drop-in, zero Aerie change):** Keycloak (dotted `claim.name` mappers), Auth0 (Actions add a namespaced claim — the canonical Hasura pattern), and self-hosted alternatives like Authentik / Zitadel.
- **Camp B — emits flat claims only (roles as a top-level or array claim, no nested object):** AWS Cognito (Lambda claim values stringify), Azure AD / Entra ID (flat app-roles/groups, plus v1/v2 + groups-overage quirks; Azure Government for gov deployments), Okta (flat value-expression claims + groups), Google Identity (minimal custom-claim support — least suitable).

The robust fix for the whole of Camp B is **not** per-IdP gymnastics — it's a small, confined Aerie enhancement to accept flat / `claims_map`-style claims (teach `extractClaims` / `getClaimsConfig` / gateway `session()` a "flat mode": `x-hasura-user-id` from `sub`, roles from `cognito:groups` / `groups` / app-roles). Hasura already handles the flat shape via `claims_map`; Aerie's own extraction is the only missing piece. That single change unlocks Cognito, Azure, Okta, and Google uniformly. (Mapper ceilings vary by IdP version/license tier, so treat the A/B split as "as typically configured.")

#### Keycloak (Camp A — fully supported today)

Mirror the vendored test realm (`e2e-tests/oauth/realm-export.json`). On a client scope attached to the `aerie` client (Authorization Code + PKCE, redirect URI `<origin>/oidc/callback`), add three protocol mappers. Keycloak's **dotted Token Claim Name** is what produces the real nested JSON object:

| Hasura claim | Keycloak mapper type | Token Claim Name | Notes |
|---|---|---|---|
| `x-hasura-user-id` | User Property (`oidc-usermodel-property-mapper`) | `https://hasura.io/jwt/claims.x-hasura-user-id` | JSON type String; add to ID **and** access token |
| `x-hasura-allowed-roles` | Client Role (`oidc-usermodel-client-role-mapper`) | `https://hasura.io/jwt/claims.x-hasura-allowed-roles` | **Multivalued** ON → emits a JSON array |
| `x-hasura-default-role` | User Attribute (`oidc-usermodel-attribute-mapper`) | `https://hasura.io/jwt/claims.x-hasura-default-role` | Reads a per-user `default_role` attribute; single-valued |

Each user gets an explicit `default_role` attribute (the deterministic-default decision described above). Auth0 is the equivalent low-effort hosted option — an Action adds the same namespaced claim.

#### AWS Cognito (Camp B — needs one of the following)

Cognito can't emit a nested JSON claim object with hyphenated keys: neither its console mappers nor its Pre-Token-Generation Lambda produce nested JSON (Lambda claim values coerce to strings). Since Aerie's `extractClaims` requires `token[namespace]` to be an object, a raw Cognito token won't satisfy the UI/gateway as-is. Options, **best-first**:

1. **Add flat-claims support to Aerie (recommended).** A small, contained follow-up to read flat claims — `x-hasura-user-id` from `sub`, `x-hasura-allowed-roles` from `cognito:groups`, `x-hasura-default-role` from a `custom:default_role` attribute. Pair with Hasura's native `claims_map` so Hasura consumes the same flat token with no Lambda:

   ```json
   {
     "type": "RS256",
     "jwk_url": "https://cognito-idp.<region>.amazonaws.com/<poolId>/.well-known/jwks.json",
     "claims_map": {
       "x-hasura-user-id":       { "path": "$.sub" },
       "x-hasura-allowed-roles": { "path": "$['cognito:groups']" },
       "x-hasura-default-role":  { "path": "$['custom:default_role']", "default": "viewer" }
     }
   }
   ```

   This is the durable fix and benefits every other Camp B IdP too. Cognito specifics: roles from `cognito:groups`; per-user default via a `custom:default_role` attribute; confirm the custom claims land on the token sent as the Bearer credential (ID vs access token depends on the Lambda trigger target).

2. **Pre-Token-Generation Lambda (V2 trigger).** Inject `x-hasura-*` claims and surface `cognito:groups` via `groupOverrideDetails`. Feasible but fiddly, and Cognito's stringification of complex values means you'll generally still lean on option 1 to deliver the shape Aerie expects. Fallback, not default.

3. **Keycloak / Auth0 identity brokering in front of Cognito.** Users authenticate against Cognito upstream; the broker federates it and re-issues an Aerie-facing token with the nested claims. Zero Aerie code change, but it means **operating a stateful IdP (service + DB) purely to reshape claims** — only worth it if you already run one or want federation for other reasons. Do not stand one up solely for this.

## Testing

```sh
# Regular e2e (JWT mode, unchanged)
npm run test:e2e

# OIDC e2e (new — requires the OIDC stack running)
npm run test:e2e:oidc:setup
npm run test:e2e:oidc
npm run test:e2e:oidc:teardown
```

The OIDC suite covers: login as admin/user/viewer, role switching, refresh, logout, multi-tab refresh coordination, tab-backgrounding refresh.

<details><summary>Manual test scenarios</summary>

Prerequisites: OIDC stack up (`npm run test:e2e:oidc:setup`), UI started with OIDC env vars sourced. Browser must support the Cookie Store API (Chrome/Edge) for automatic token refresh.

**Normal operation (happy path)**

1. Navigate to the app → redirected to IdP login
2. Log in with valid credentials → redirected back to the app
3. Verify user is authenticated and can access protected pages
5. Open DevTools Network tab
6. Wait for the next refresh (~10s before access-token expiry) → `/oidc/refresh` request appears
7. Verify WebSocket connection restarts cleanly (look for a new GraphQL WS with fresh `connection_init`)
8. Verify no "Reconnecting…" banner appears during normal refresh
9. Switch user role via dropdown → WS restarts, data reloads

**Offline resilience**

1. While logged in with valid tokens, enable "Offline" in DevTools Network
2. Wait for next scheduled token refresh
3. Observe token refresh fails in console
4. Disable "Offline" (back online)
5. Verify refresh retries automatically and succeeds within ~5s
6. Verify WS reconnects and subscriptions resume
7. Verify no data loss or UI errors

**Role switching**

1. Log in and open a page with subscription data
2. Watch WS messages in DevTools
3. Switch role via the Nav dropdown
4. Verify WS closes and reopens with new `x-hasura-role` in headers
5. Verify subscriptions re-register and data updates for the new role's permissions

**Logout flow**

1. Click logout
2. Verify cookies are cleared (DevTools → Application tab)
3. Verify the app redirects to the IdP logout endpoint
4. Verify the IdP redirects back, then app redirects to `/login`
5. Check console + server logs — should be clean

**Long-running session**

1. Log in and leave the app running for >1 hour (assumes refresh-token TTL > 1h)
2. Periodically check Network for refresh activity
3. Verify WS stays healthy across refreshes
4. Interact with the app (create/edit plans, etc.) to verify subscriptions
5. Verify no unexpected logouts or connection errors

**HMR during development (dev only)**

1. Log in and navigate to a subscription-heavy page
2. Make a code change to trigger HMR
3. Verify subscriptions reconnect seamlessly if tokens are fresh
4. If tokens expired during HMR idle time: verify auto-recovery and that subscriptions reconnect once refresh runs

</details>

## Big decisions and where the complexity is

- **OIDC lives in the UI; user provisioning lives in the gateway.** Why each piece is where it is — see "Why OIDC lives in the UI rather than the gateway" above. Short version: PKCE cookies and refresh trigger have to be in the UI because of how the browser hits the IdP and Hasura; user-row provisioning fits the gateway because that's where every other auth mode writes the row today.

- **WebSocket lifecycle + token refresh** — non-obvious behavior worth a closer read, collapsed because it's longest:

  <details><summary>WS Lifecycle & Token Refresh details</summary>

  Hasura validates JWT not only at `connection_init` but also continuously monitors token expiration, closing WebSocket connections when JWTs expire (observed in Hasura logs: `"Could not verify JWT: JWTExpired"`, close codes 1006/4400).

  Implementation:

  - **Proactive WebSocket restart** on cookie refresh prevents abrupt disconnects when the access token rolls over. The close uses a named code+reason pair (`INTENTIONAL_RESTART_CODE = 4999`, `INTENTIONAL_RESTART_REASON = "Client Restart"`) so the `closed` handler can recognize *our* restart and suppress the transient `'reconnecting'` state — otherwise the banner would flash on every refresh. Gating on both code AND reason prevents a foreign close at the same code from accidentally swallowing the banner. Trade-off: if our reconnect ever hangs on a network/server issue, the banner won't surface until graphql-ws's own `connectionAckWaitTimeout` fires (~15s) and a non-4999 close follows.
  - **Automatic token refresh** via the Cookie Store API; the refresh subscriber listens on `accessTokenDecoded` (object, never deduped) rather than the numeric `delay` store (which Svelte primitive-deduped, causing refreshes to silently stop after identical-delay cycles)
  - **Hybrid auto-recovery**: subscriptions recover from connection errors using two strategies — a `connectionState` listener (fast, ~100ms when graphql-ws auto-reconnects) and a 5s fallback timer (kicks graphql-ws out of lazy mode when all subscriptions are terminated)
  - **Logout reason survives the IdP roundtrip** in OIDC mode. `/oidc/refresh` returning 401 → `logout('Session expired - please log in again')` → `/oidc/logout?reason=...` → server stashes the reason in an `httpOnly` `logoutReason` cookie (60s TTL) → IdP `end_session_endpoint` → IdP redirects back to origin (query params stripped) → `+layout.server.ts` consumes and deletes the cookie, appending `&reason=...` to the `/login` redirect → existing `/login` reason handling surfaces it to the user.
  - **Offline resilience**: token refresh retries every 5s on network failure until successful
  - **HMR resilience**: cookie-store listeners and WS client state persist across Hot Module Replacement during development
  - **WS auth detection** in `on.error` matches both `JWTExpired` and `JWSError` (signature-invalid tokens, e.g., from IdP key rotation or tampering)

  </details>

- **Refresh trigger lives in the client.** The "gateway" can't intercept Hasura HTTP/WS calls (the UI hits Hasura directly, the gateway isn't a network gateway), so the refresh trigger has to live in the browser. setTimeout-based, with refresh-on-401 fallback in the WS error handler.

- **Auth-mode env coupling.** Hasura's `HASURA_GRAPHQL_JWT_SECRET` is process-global; one Hasura instance can't validate both HS256 (gateway-signed) and RS256 (Keycloak-signed) tokens simultaneously on v2.12.1. CI runs two phases: the regular suite with HS256, then OIDC with RS256.

- **Each user's "default role" comes from a setting on their IdP profile, not from a guess.** A user can have multiple roles (e.g., an admin who also has `user` and `viewer`). When they log in, the UI has to pick *one* to start in. The previous Keycloak config asked Keycloak to pick a role from the user's list, and Keycloak picked arbitrarily — so an admin could land in viewer-mode unpredictably.

  Now each IdP user has an explicit `default_role` attribute on their profile (e.g., `"default_role": ["aerie_admin"]`), and that attribute is what gets stamped into the JWT. Two practical benefits:

  1. **Deterministic** — same user gets the same starting role every login.
  2. **Mission-configurable per user** — a mission can default senior planners to `viewer` for safer day-to-day use even though those users have higher privileges they can opt into via the dropdown. We don't hardcode an "admin > user > viewer" priority anywhere in the UI.

  Configured in `e2e-tests/oauth/realm-export.json` (test realm); each production deployment configures the same on their own IdP.

## Open questions

- **Unify cookie storage across auth modes (descoped from this PR).** JWT/SSO modes store the user's data in a base64-encoded `user` cookie (`{ id, token }`); OIDC mode stores the access token directly as a plain `accessToken` cookie. The shared `getToken()` in `src/stores/gqlClient.ts:81-102` has to branch on which shape it sees, and similar mode-specific branching exists in the hooks dispatcher and gateway-side cookie handling. Long-term, unifying on the OIDC-style (`accessToken` cookie + derive `id` from JWT claim) would collapse a lot of duplication. Out of scope here because it's cross-repo (gateway constructs the `user` cookie today) and the refactor logically pulls in SSO/CAM's cookie handling too — both modes whose auth we don't want to risk regressing in this PR. Worth its own coordinated PR after this lands.
- **HTTPS-only cookies in production.** `src/lib/server/oidc.ts` sets `secure: !dev` on auth cookies — works in dev (HTTP) and production over HTTPS, but breaks if a production deploy ever runs behind HTTP (e.g., internal network without TLS). Worth a follow-up to thread the request URL in and compute `secure = url.protocol === 'https:'`.
- **Cosmetic trailing duplicate refresh + logout.** The refresh `setTimeout` isn't cancelled when `logout()` is called, so it can fire once during the navigation to `/oidc/logout`, producing an extra `/oidc/refresh` + `/oidc/logout` pair in logs. No functional impact. Optional fix: `cancelScheduledRefresh()` called from `logout()`.
- **Plan-revision subscription destructure-of-null** (pre-existing, surfaced more often with frequent WS restarts): `src/stores/plan.ts:111` destructures `revision` from `data` without null-guarding. Out of scope for this PR, separate follow-up.
- **Hasura upgrade path** (v2.12 → v2.18+) would enable `HASURA_GRAPHQL_JWT_SECRETS` (plural) so one Hasura could validate both HS256 and RS256 simultaneously, eliminating the CI two-phase split.

## TODOs

- [ ] Make `secure` flag depend on request protocol, not just `dev` mode (production-HTTP deploy correctness)
- [ ] `cancelScheduledRefresh()` in `logout()` to silence the cosmetic duplicate refresh + logout pair
- [ ] Decide whether to fix `plan.ts:111` destructure-of-null here or as a separate PR (it's pre-existing, not OIDC-introduced)
- [ ] Document the local OIDC dev story in `docs/TESTING.md` (or rely on this PR description; either way, devs need to know about `test:e2e:oidc:setup`)

jmorton committed:
Add OIDC authentication support with Arctic library

Implement complete OIDC authentication flow including:
- Server-side OIDC client using Arctic library with PKCE support
- Login, callback, logout, and token refresh endpoints
- Updated hooks.server.ts to handle OIDC authentication mode
- Modified subscribable stores and effects for token management
- Request utilities updated for authenticated API calls

jmorton committed:
Replace PageData.user with reactive userStore across application

Migrate from passing user through PageData to using a centralized
userStore for authentication state. This change:
- Removes user parameter threading through page components
- Updates all stores to access user from centralized auth store
- Refactors route layouts to use reactive auth state
- Removes unnecessary +page.ts files that only passed user data
- Enables role changes to propagate without full page reload

# Conflicts:
#	src/components/plan/PlanMergeReview.svelte
#	src/routes/+layout.server.ts
#	src/routes/+layout.svelte
#	src/routes/constraints/+layout.svelte
#	src/routes/constraints/+page.svelte
#	src/routes/constraints/edit/[id]/+page.svelte
#	src/routes/constraints/new/+page.svelte
#	src/routes/dictionaries/+page.svelte
#	src/routes/expansion/+layout.svelte
#	src/routes/expansion/rules/+page.svelte
#	src/routes/expansion/rules/edit/[id]/+page.svelte
#	src/routes/expansion/rules/new/+page.svelte
#	src/routes/expansion/runs/+page.svelte
#	src/routes/expansion/sets/+page.svelte
#	src/routes/expansion/sets/new/+page.svelte
#	src/routes/external-sources/+layout.svelte
#	src/routes/external-sources/sources/+page.svelte
#	src/routes/external-sources/types/+page.svelte
#	src/routes/models/+layout.svelte
#	src/routes/models/+page.svelte
#	src/routes/models/[id]/+page.svelte
#	src/routes/parcels/+layout.svelte
#	src/routes/parcels/+page.svelte
#	src/routes/parcels/edit/[id]/+page.svelte
#	src/routes/parcels/new/+page.svelte
#	src/routes/plans/+page.svelte
#	src/routes/plans/[id]/+page.svelte
#	src/routes/plans/[id]/merge/+page.svelte
#	src/routes/scheduling/+layout.svelte
#	src/routes/scheduling/+page.svelte
#	src/routes/scheduling/conditions/edit/[id]/+page.svelte
#	src/routes/scheduling/conditions/new/+page.svelte
#	src/routes/scheduling/goals/edit/[id]/+page.svelte
#	src/routes/scheduling/goals/new/+page.svelte
#	src/routes/sequence-templates/+layout.svelte
#	src/routes/sequence-templates/+page.svelte
#	src/routes/tags/+page.svelte
#	src/routes/workspaces/+layout.svelte
#	src/routes/workspaces/+page.svelte
#	src/routes/workspaces/[workspaceId]/+layout@.svelte
#	src/routes/workspaces/[workspaceId]/actions/+layout@.svelte
#	src/routes/workspaces/[workspaceId]/actions/runs/[runId]/+layout@.svelte
#	src/routes/workspaces/[workspaceId]/actions/runs/[runId]/+page.svelte
#	src/stores/sequencing.ts
#	src/stores/tags.ts

jmorton committed:
Add server-side rule enforcement infrastructure

Create rule.ts module for enforcing authentication requirements
at the route level, enabling consistent access control across
the application.

jmorton committed:
Add E2E tests for OIDC authentication flow

Implement Playwright tests for OIDC authentication including:
- OIDC fixture for handling auth flows in tests
- Login/logout test scenarios
- Token refresh verification
- Updated helpers and AppNav fixture for OIDC support

jmorton committed:
Update non-OIDC auth routes to use event-based cookie handling

Align legacy auth routes (login, logout, changeRole) with OIDC
implementation by using SvelteKit's event-based cookie API
instead of manual header manipulation. Reduces code duplication
and improves consistency.

jmorton committed:
Fix race condition in OIDC well-known configuration fetch

The Client singleton was initiating a fetch for the well-known
configuration but not awaiting it, causing endpoint values to be
undefined when the constructor completed before the fetch.

Changes:
- Convert Client.instance to async getter returning Promise<Client>
- Move initialization to async init() method that awaits well-known fetch
- Update all call sites to await Client.instance
- Uncomment OIDC_CLIENT_SECRET env var to fix type error

jmorton committed:
Add OIDC nonce parameter for replay attack protection

The nonce parameter binds the ID token to a specific authentication
request, preventing attackers from reusing previously issued tokens.

Changes:
- Add generateNonce() function using crypto.randomBytes
- Include nonce in authorization URL via createAuthorizationURLWithPKCE()
- Store nonce in httpOnly cookie during login
- Add verifyNonce() to validate ID token nonce claim on callback
- Update callback check() to require nonce presence

jmorton committed:
Add audience claim validation to JWT verification

Validate the 'aud' claim in tokens to prevent confused deputy attacks
where tokens issued for other applications could be accepted.

Set OIDC_AUDIENCE env var to enable validation (recommended for production).

jmorton committed:
Validate audience only on ID token per OIDC spec

- Split JWT verification options: BASE_VERIFY_OPTS (no audience) for
  access tokens, ID_TOKEN_VERIFY_OPTS (with audience) for ID tokens
- Access tokens are treated as opaque per OIDC spec - audience
  validation is the resource server's responsibility
- ID tokens require audience validation to prevent confused deputy attacks
- Also fixes secure cookie flag for local development (secure: !dev)

jmorton committed:
Fix open redirect vulnerability in back parameter

Validate the `back` query parameter in the OIDC login route to prevent
open redirect attacks. Only allow relative paths starting with '/'
but not '//' (protocol-relative URLs).

Rejected examples: 'https://evil.com', '//evil.com', 'javascript:alert(1)'

jmorton committed:
Add secure flag to all authentication cookies

Ensures auth cookies are only transmitted over HTTPS in production.
Uses `secure: !dev` pattern to allow HTTP in local development.

Updated files:
- oidc.ts: accessToken, idToken, refreshToken
- hooks.server.ts: activeRole (OIDC and SSO handlers)
- oidc/login/+page.server.ts: back cookie
- auth/login/+server.ts: activeRole, user
- auth/changeRole/+server.ts: activeRole

jmorton committed:
Add SameSite=Lax to all authentication cookies

Using 'lax' instead of 'strict' because:
- OIDC flow requires cookies to persist across the redirect back from Keycloak
- 'lax' allows cookies on top-level GET navigations (safe for redirects)
- 'lax' still blocks cross-site POST requests (CSRF protection)

Note: SSO handler keeps sameSite='none' for cross-site SSO compatibility.

jmorton committed:
Add configurable JWT signing algorithms via OIDC_ALGORITHMS env var

Allows operators to specify supported JWT signing algorithms via
space-separated OIDC_ALGORITHMS environment variable (e.g., "RS256 RS384").
Defaults to RS256, the most common algorithm for OIDC providers.

jmorton committed:
Remove sensitive data from OIDC logs

- Remove token values from user registration logs (only log username)
- Remove event object from cookie change log (contained token values)
- Remove token from error messages in callback
- Log only error messages, not full error objects that may contain tokens
- Downgrade verbose logs to console.debug

jmorton committed:
Standardize OIDC logging and fix additional sensitive data leaks

- Change informational logs from console.log to console.debug for consistency
- Remove response body from refresh error (may contain sensitive details)
- Only log error messages, not full error objects in scheduled refresh
- Remove timeout ID from log output (not useful for debugging)
- Improve log message clarity ("Scheduling token refresh" vs "Delay changed")

jmorton committed:
Add Content Security Policy headers (report-only mode)

Adds CSP and other security headers to all responses:
- Content-Security-Policy-Report-Only: monitors violations without blocking
- X-Content-Type-Options: nosniff
- X-Frame-Options: DENY
- Referrer-Policy: strict-origin-when-cross-origin

CSP directives configured for:
- Scripts: self + unsafe-inline + unsafe-eval (Monaco editor requires these)
- Styles: self + unsafe-inline (Svelte scoped styles)
- Connect: self + Hasura + Gateway + Action + Workspace URLs
- Workers: self + blob (Monaco editor)
- Images/Fonts: self + data + blob

Change header to 'Content-Security-Policy' to enforce after testing.

jmorton committed:
Verify ID token before using as logout hint

Add verifyIdToken() function to validate ID tokens with full OIDC-compliant
checks (signature, issuer, expiration, audience) before passing to IdP.

If verification fails during logout, proceed without the id_token_hint
parameter rather than sending an unverified token.

jmorton committed:
Add configurable JWT claim paths for OIDC

Add environment variables to configure JWT claim namespace and paths:
- OIDC_CLAIMS_NAMESPACE (server) / PUBLIC_OIDC_CLAIMS_NAMESPACE (client)
- OIDC_CLAIMS_USER_ID / PUBLIC_OIDC_CLAIMS_USER_ID
- OIDC_CLAIMS_ALLOWED_ROLES / PUBLIC_OIDC_CLAIMS_ALLOWED_ROLES
- OIDC_CLAIMS_DEFAULT_ROLE / PUBLIC_OIDC_CLAIMS_DEFAULT_ROLE

Defaults to Hasura's standard claim structure:
  https://hasura.io/jwt/claims -> x-hasura-user-id, x-hasura-allowed-roles, x-hasura-default-role

Add extractClaims() helper functions in both oidc.ts (server) and auth.ts (client)
to centralize claim extraction with proper validation.

IMPORTANT: These settings must match:
- Hasura's HASURA_GRAPHQL_JWT_SECRET claims_map
- Aerie Gateway's JWT parsing logic
- Your IdP's token mapper configuration

jmorton committed:
Use dynamically loaded env vars for OIDC related values set at launch / runtime.

AaronPlave commented:
@jmorton @pranav-super We should tag up about this PR in relation to my new changes in https://github.com/NASA-AMMOS/aerie-ui/pull/1741 where I've added socket consolidation + user store defined in a client-side store passed around via svelte context.

duranb commented on src/lib/server/rule.ts:
```diff
@@ -0,0 +1,10 @@
+  return !!u;
+};
+
+export const userIsAdmin: Rule = (u: User | null) => {
```
This function is already defined in `permissions.ts`

duranb commented on src/lib/stores/auth.ts:
```diff
@@ -0,0 +1,7 @@
+import { type Client } from 'graphql-ws';
```
Can we move these store files to the already existing `/stores` directory? Is there a need to keep them separate?

duranb commented on src/utilities/auth.ts:
```diff
@@ -0,0 +1,119 @@
+  };
+}
+
+export function goToLogin() {
```
Is this function being used?

duranb commented on src/utilities/auth.ts:
```diff
@@ -0,0 +1,119 @@
+
+export function goToLogin() {
+  if (env.PUBLIC_AUTH_OIDC_ENABLED === 'true') {
+    document.location.href = '/oidc/login';
```
If possible, please use a more svelte-like convention of redirecting. [This](https://svelte.dev/tutorial/kit/redirects) might be more appropriate?

duranb commented on src/utilities/auth.ts:
```diff
@@ -0,0 +1,119 @@
+ */
+const CLAIMS_CONFIG = {
+  namespace: env.PUBLIC_OIDC_CLAIMS_NAMESPACE || 'https://hasura.io/jwt/claims',
+  userId: env.PUBLIC_OIDC_CLAIMS_USER_ID || 'x-hasura-user-id',
```
Is this for the actual user's `userId` or a way to know under what header name to look for the user's `userId`? If it is the latter, can we rename these to be less ambiguous?

duranb commented on src/utilities/auth.ts:
```diff
@@ -0,0 +1,119 @@
+ * Falls back to Hasura's standard claim namespace.
+ */
+const CLAIMS_CONFIG = {
+  namespace: env.PUBLIC_OIDC_CLAIMS_NAMESPACE || 'https://hasura.io/jwt/claims',
```
Can we potentially just generalize the env var names so that they can potentially be used by any auth paradigm? Something like `PUBLIC_AUTH_CLAIMS_NAMESPACE`, `PUBLIC_AUTH_CLAIMS_USER_ID`, etc.

duranb commented on src/utilities/auth.ts:
```diff
@@ -0,0 +1,119 @@
+    throw new Error(`JWT missing claims namespace: ${CLAIMS_CONFIG.namespace}`);
+  }
+
+  const userId = namespace[CLAIMS_CONFIG.userId] as string;
```
I don't think you need to typecast these to their final type here. At this point, I think it's still fine to consider them `unknown`. The following checks ensure that what is returned is what we expect.

AaronPlave committed:
Fix WebSocket lifecycle and token refresh for OIDC

Key changes:
- Add proactive WebSocket restart on token refresh (Hasura monitors JWT expiration and kills connections)
- Implement hybrid auto-recovery for subscription errors (connectionState listener + fallback timer)
- Add token refresh retry on failure (offline resilience)
- Fix HMR resilience for cookie store listeners
- Prevent token refresh on login page
- Fix logout to use window.location.href instead of goto() for server-only routes
- Remove noisy auth error logging during normal logout flow
- Add getCookieValue utility for reading cookies

AaronPlave committed:
Linting, fixes, and small cleanup

AaronPlave committed:
Handle expired refresh tokens gracefully instead of infinite retry loop

- Return 401 from /oidc/refresh instead of throwing on expired tokens
- Catch refresh failures on page load, clear cookies, redirect to login
- Fix delay=0 being falsy so expired tokens trigger immediate refresh
- On 401, log out immediately instead of retrying
- Send raw id_token as logout hint (IdPs accept expired tokens for this)
- Handle missing id_token in multi-tab logout scenarios

AaronPlave committed:
Merge remote-tracking branch 'origin/develop' into feature/oidc-support

# Conflicts:
#	src/routes/workspaces/[workspaceId]/actions/+layout.ts
#	src/routes/workspaces/[workspaceId]/actions/+page.svelte
#	src/routes/workspaces/[workspaceId]/actions/runs/[runId]/+layout.ts

AaronPlave committed:
Revert reqHasura auto-logout, dedupe OIDC claims helper, anchor non-protected paths

AaronPlave committed:
Add OIDC role-switch, multi-tab, and tab-backgrounding e2e tests and clean up existing fixture

AaronPlave committed:
Surface auto-logout reason on /login after OIDC roundtrip and suppress reconnecting banner on intentional WS restarts

AaronPlave committed:
Move OIDC user provisioning to gateway and add OIDC e2e test stack

AaronPlave committed:
Harden OIDC role-switch e2e and refresh stale test comments

AaronPlave committed:
Lint fixes

AaronPlave committed:
Merge branch 'develop' into feature/oidc-support

AaronPlave committed:
test fix

AaronPlave committed:
add backend-service smoke tests to the OIDC e2e suite

AaronPlave committed:
Split OIDC tests into their own runner

AaronPlave committed:
Merge remote-tracking branch 'origin/develop' into feature/oidc-support

# Conflicts:
#	src/utilities/login.ts

AaronPlave committed:
test: dedupe snapshot helper onto develop's createPlanSnapshot

The develop merge brought in AerieApi.createPlanSnapshot, which duplicated
the createSnapshot helper this branch added (both wrap CREATE_PLAN_SNAPSHOT).
Drop the branch's createSnapshot and point the OIDC snapshot smoke test at
createPlanSnapshot.

AaronPlave committed:
fix(e2e): skip templating preview server on OIDC runs

The OIDC CI job intentionally skips `npm run build` (the OIDC Playwright
project serves the UI via `npm run dev`). But the second webServer entry
(the command-expansion 'templating' variant on port 3001) always ran
`npm run preview`, which needs build output that isn't there — so it failed
to start and aborted the entire OIDC run before any test executed. OIDC
tests don't use port 3001, so only stand up that preview server on
non-OIDC runs.

AaronPlave committed:
fix(oidc): treat empty OIDC_ALGORITHMS as default RS256

getSupportedAlgorithms did `env.OIDC_ALGORITHMS?.split(' ') || ['RS256']`.
But `''.split(' ')` returns `['']` — a truthy one-element array — so the
`|| ['RS256']` fallback never fired when OIDC_ALGORITHMS was set-but-empty
(as it is in the committed .env). verify() then ran with algorithms: [''],
and every Keycloak RS256 token was rejected with 'invalid algorithm',
failing every OIDC login callback in CI. Trim, drop empties, and fall back
to RS256 when nothing valid remains.

AaronPlave committed:
test(oidc): warm up dev server before OIDC suite to de-flake first login

The OIDC suite runs the UI via `npm run dev` (preview mode drops secure
cookies over http). Vite dev lazy-compiles routes on first request, so the
first test's goto('/plans') ate a cold compile that raced the OIDC redirect
chain and aborted the navigation (net::ERR_ABORTED), timing out at 30s and
flaking the first login every run. Add a setup project that primes /login
and /plans before the suite, so tests start against warm modules.
